atomic_reactor/plugins/pre_resolve_module_compose.py

```diff
@@ -0,0 +1,266 @@
+                         burst_retry=1,
+                         burst_length=30,
+                         slow_retry=10,
+                         timeout=300):
+        self.log.debug("Getting compose information for information for compose_id={}"
+                       .format(compose_id))
+        url = self.url + 'composes/' + str(compose_id)
+        headers = self._auth_headers()
+        start_time = time.time()
+        while True:
+            response = requests.get(url, headers=headers)
+            response.raise_for_status()
+            response_json = response.json()
+
+            if response_json['state_name'] not in ['wait', 'generating']:
+                self.log.debug("Retrieved compose information for compose_id={}: {!r}"
+                               .format(compose_id, response_json))
+                return response_json
+
+            elapsed = time.time() - start_time
+            if elapsed > timeout:
+                raise RuntimeError("Retrieving %s timed out after %s seconds" %
+                                   (url, timeout))
+            else:
+                self.log.debug("Retrying request compose_id={}, elapsed_time={}"
+                               .format(compose_id, elapsed))
+
+                if elapsed > burst_length:
+                    time.sleep(slow_retry)
+                else:
+                    time.sleep(burst_retry)
+
+
+class ResolveModuleComposePlugin(PreBuildPlugin):
+    key = "resolve_module_compose"
+    is_allowed_to_fail = False
+
+    def __init__(self, tasker, workflow,
+                 module_name, module_stream, module_version=None,
+                 compose_id=None,
+                 odcs_url=None, odcs_insecure=False,
+                 odcs_openidc_secret_path=None,
+                 pdc_url=None, pdc_insecure=False,
+                 compose_url=None):
+        """
+        constructor
+
+        :param tasker: DockerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param module_name: Module name to look up in PDC
+        :param module_stream: Module stream to look up in PDC
+        :param module_version: Module version to look up in PDC (optional)
+        :param compose_id: ID of compose in ODCS (optional - will only be set for workers)
+        :param odcs_url: URL of ODCS (On Demand Compose Service)
+        :param odcs_insecure: If True, don't check SSL certificates for `odcs_url`
+        :param odcs_openidc_secret_path: directory to look in for a `token` file (optional)
+        :param pdc_url: URL of PDC (Product Definition center))
+        :param pdc_insecure: If True, don't check SSL certificates for `pdc_url`
+        :
+        """
+        # call parent constructor
+        super(ResolveModuleComposePlugin, self).__init__(tasker, workflow)
+
+        if not pdc_url:
+            raise RuntimeError("pdc_url is required")
+        if not odcs_url:
+            raise RuntimeError("odcs_url is required")
```

Comment:
Grouping is certainly one consideration. Another consideration that may have been in my head is that  python2, you get very unclear messages when trying to pass non-kwargs as kwargs goes wrong (foo() takes at least 4 arguments (6 given) - python3 is better. I think I'll keep it this way since I do like having the parameters grouped.